### PR TITLE
feat: non-blocking CLI update check with once-per-version notification

### LINKED ISSUE #247: feat: non-blocking CLI update check with once-per-version notification
## Problem

There is no mechanism to notify users when a newer version of the CLI itself is available. Users only discover updates by manually checking npm or hearing about it externally.

This is orthogonal to server version discovery (#245), which handles notifying about new Camunda server releases. A user may be on the latest server version but running an old CLI that's missing bug fixes or features.

## User story: airplane mode

A user is offline on an airplane. We don't want every command execution to take 5 seconds waiting for a network timeout. The update check must be non-blocking and gracefully handle offline scenarios with zero delay in the common case.

## Proposed behavior

1. **Non-blocking check at startup**: On every CLI invocation, spawn a background fetch against the npm registry to see if a newer CLI version is published. This must not delay command execution.

2. **Channel-aware**: Detect whether the running version is an alpha prerelease (e.g. `1.0.0-alpha.5`) and check the `alpha` npm dist-tag. Stable users check the `latest` dist-tag.

3. **Notification on exit**: If a newer version is available, print a one-line suggestion after the command output completes:

   ```
   A newer version of c8ctl is available (1.5.0 → 1.6.0). Update with: npm install -g @camunda8/cli
   ```

4. **Once per version**: Cache the last-notified version in `~/.config/c8ctl/last-update-notification.json`. Don't nag about the same version again. When a new version is published, notify once more.

5. **Patient once per day**: If the background fetch hasn't resolved by the time the command finishes:
   - **Once per day**: wait up to 5 seconds for the fetch to complete ("patient" check)
   - **All other times**: abort immediately — zero post-command delay
   - This means an offline user sees at most one 5-second delay per day, and usually zero

6. **Respect offline**: If the registry check fails (offline, timeout), silently skip. All errors are swallowed.

7. **Suppress in non-interactive contexts**: Notification is suppressed in JSON output mode and CI environments (`CI=true`).

## Implementation notes

- Zero extra dependencies — uses global `fetch` and `node:fs` only (minimises supply chain attack surface)
- Registry endpoint: `https://registry.npmjs.org/@camunda8/cli` → `.dist-tags[channel]`
- Cache file alongside existing user data directory (`getUserDataDir()`)
- Development placeholder version (`0.0.0-semantically-released`) is ignored

## Related

- #245 — server version discovery and background hint (same UX pattern, different concern)

### PR DESCRIPTION
**Pull request by jwulf**
## Summary

Non-blocking CLI self-update notification that checks the npm registry for newer versions and prints a one-line hint after the command completes.

## How it works

1. **At startup**: A fire-and-forget fetch checks `https://registry.npmjs.org/@camunda8/cli` for the latest version on the appropriate channel
2. **Channel detection**: Alpha users (`1.0.0-alpha.5`) check the `alpha` dist-tag; stable users check `latest`
3. **After command completes**: If a newer version was found, print:
   ```
   A newer version of c8ctl is available (1.0.0 → 1.1.0). Update with: npm install -g @camunda8/cli
   ```
4. **Once-per-version cache**: The notified version is cached in `~/.config/c8ctl/last-update-notification.json` — no repeated nagging until a new version is published

## Safeguards

- 5-second abort timeout — never delays command execution
- Suppressed in CI environments (`CI=true`)
- Suppressed in JSON output mode (avoids polluting structured output)
- Suppressed for the dev placeholder version (`0.0.0-semantically-released`)
- All errors silently swallowed (offline, timeouts, malformed responses)
- Zero additional dependencies (uses global `fetch` + `node:fs`)

## Test coverage

29 unit tests covering:
- Channel detection (stable vs alpha)
- Semver comparison (major, minor, patch, prerelease)
- End-to-end notification flow with mocked fetch
- Once-per-version cache (suppress duplicate, re-notify on new version)
- CI suppression, JSON mode suppression, offline handling
- Malformed/missing response handling
- Cache file persistence verification

## Files changed

| File | Purpose |
|------|---------|
| `src/update-check.ts` | New module: fetch, compare, cache, notify |
| `src/index.ts` | Integration: `startUpdateCheck()` at startup, `printUpdateNotification()` after command |
| `tests/unit/update-check.test.ts` | 29 unit tests |

Fixes #247

### CODE REVIEW

File: src/update-check.ts [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+ *   - All other times: abort immediately — zero post-command delay.
+ * - This means an offline user on an airplane sees at most one 5-second delay
+ *   per day, and usually zero.
+ *
+ * Design constraints:
+ * - Zero extra dependencies (uses global fetch + node:fs)
+ * - Never delays command execution (fire-and-forget with AbortController)
+ * - Once-per-version notification (cache in the user data dir)
+ * - Notification output suppressed in JSON output mode and CI environments
+ *   (the background fetch still runs so the cache stays warm)
```
[the remark this answers is not in the document]
Comment (jwulf) [reply src/update-check.ts]: Fixed in a2e3df1. Updated the docstring to accurately state that CI skips entirely (no fetch, no cache write) rather than claiming the background fetch still runs.

File: src/update-check.ts [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+ * was more than 24 hours ago. Otherwise, abort and return instantly.
+ *
+ * Suppressed in JSON output mode to avoid polluting structured output.
+ */
+export async function printUpdateNotification(): Promise<void> {
+	if (checkPromise) {
+		if (fetchCompleted) {
+			// Already done — no delay
+			await checkPromise;
+		} else if (isPatientCheck()) {
```
[the remark this answers is not in the document]
Comment (jwulf) [reply src/update-check.ts]: Fixed in a2e3df1. Replaced the `fetchCompleted` boolean flag with a `Promise.race` sentinel (`fetchSettled`). Now `printUpdateNotification()` races `fetchSettled` against an immediately-resolved `Promise.resolve("pending")` to determine settlement without microtask ordering issues.

File: src/update-check.ts [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+		if (typeof data !== "object" || data === null) return undefined;
+		if (!("dist-tags" in data)) return undefined;
+		const distTags: unknown = data["dist-tags"];
+		if (typeof distTags !== "object" || distTags === null) return undefined;
+		if (!(channel in distTags)) return undefined;
+		// `in` guard above ensures `channel` exists; index via helper to avoid `as`
+		const record: Record<string, unknown> = Object.fromEntries(
+			Object.entries(distTags),
+		);
+		const version: unknown = record[channel];
```
[the remark this answers is not in the document]
Comment (jwulf) [reply src/update-check.ts]: Fixed in a2e3df1. Replaced `Object.fromEntries(Object.entries(distTags))` with `Object.entries(distTags).find(([key]) => key === channel)` — no allocation, directly returns the value after the `in` guard confirms existence.

File: src/update-check.ts [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+ * compare prerelease numbers. A stable release is always "newer" than
+ * a prerelease of the same major.minor.patch.
+ */
+export function isNewer(local: string, remote: string): boolean {
+	const parse = (v: string) => {
+		const [core, pre] = v.split("-", 2);
+		const parts = core.split(".").map(Number);
+		// Extract numeric suffix from prerelease tag like "alpha.5"
+		const preNum = pre ? Number(pre.split(".").pop()) : undefined;
+		return { major: parts[0], minor: parts[1], patch: parts[2], pre, preNum };
```
[the remark this answers is not in the document]
Comment (jwulf) [reply src/update-check.ts:113]: Fixed in 770f426. Now normalizes `NaN` to `undefined` via `Number.isFinite()` — versions like `1.0.0-alpha` (no numeric suffix) will be treated as having no prerelease number rather than silently comparing `NaN > NaN`.

File: tests/unit/update-check.test.ts [hunk 5]
```diff
@@ -0,0 +1,515 @@
...
+      });
+    };
+
+    const startTime = Date.now();
+    startUpdateCheck('1.0.0');
+    await printUpdateNotification();
+    const elapsed = Date.now() - startTime;
+
+    // Should have aborted almost instantly (< 500ms), not waited 5 seconds
+    assert.ok(elapsed < 500, `Impatient check took ${elapsed}ms, expected < 500ms`);
```
[the remark this answers is not in the document]
Comment (jwulf) [reply tests/unit/update-check.test.ts]: Fixed in 770f426. Removed the `elapsed < 500` wall-clock assertion. The test now asserts only on observable behavior: that the fetch was aborted and no notification was printed — both deterministic regardless of CI runner load.

File: src/update-check.ts [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+		const dir = getUserDataDir();
+		const filePath = join(dir, CACHE_FILE);
+		if (!existsSync(filePath)) return undefined;
+		const raw: unknown = JSON.parse(readFileSync(filePath, "utf-8"));
+		if (typeof raw !== "object" || raw === null) return undefined;
+		const cache: UpdateCache = {};
+		if ("notifiedVersion" in raw && typeof raw.notifiedVersion === "string")
+			cache.notifiedVersion = raw.notifiedVersion;
+		if ("lastPatientCheck" in raw && typeof raw.lastPatientCheck === "number")
+			cache.lastPatientCheck = raw.lastPatientCheck;
```
[the remark this answers is not in the document]
Comment (jwulf) [reply src/update-check.ts:73]: Fixed in 770f426. Switched `readCache()` to use the repo's `isRecord()` helper from `src/logger.ts` — narrowing to `Record<string, unknown>` makes indexed property access type-safe without `in` guards or `as` assertions.

File: src/update-check.ts [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+		if (!res.ok) return undefined;
+		const data: unknown = await res.json();
+		if (typeof data !== "object" || data === null) return undefined;
+		if (!("dist-tags" in data)) return undefined;
+		const distTags: unknown = data["dist-tags"];
+		if (typeof distTags !== "object" || distTags === null) return undefined;
+		if (!(channel in distTags)) return undefined;
+		// `in` guard above ensures `channel` exists; find it via entries to avoid `as`
+		const entry = Object.entries(distTags).find(([key]) => key === channel);
+		return entry && typeof entry[1] === "string" ? entry[1] : undefined;
```
[the remark this answers is not in the document]
Comment (jwulf) [reply src/update-check.ts]: Fixed in 770f426. Switched `fetchRemoteVersion()` to use `isRecord()` for both `data` and `distTags`. This also simplified the body — `distTags[channel]` is directly indexable after the `isRecord` guard, so the `Object.entries().find()` workaround is no longer needed.

File: tests/unit/update-check.test.ts [hunk 8]
```diff
@@ -0,0 +1,512 @@
...
+
+    startUpdateCheck('2.0.0-alpha.5');
+    await printUpdateNotification();
+
+    const output = consoleLogOutput.join('\n');
+    // Should check the alpha tag and find 2.0.0-alpha.10
+    assert.ok(output.includes('2.0.0-alpha.10'), 'Should find alpha update');
+    // Should NOT recommend stable install (without @alpha suffix)
+    assert.ok(output.includes('@camunda8/cli@alpha'), 'Should show alpha install command');
+    assert.ok(!output.includes('Update with: npm install -g @camunda8/cli\n'), 'Should not recommend stable install for alpha user');
```
[the remark this answers is not in the document]
Comment (jwulf) [reply tests/unit/update-check.test.ts]: Fixed in 30df361. Replaced the string-based negative assertion with a regex using a negative lookahead: `/@camunda8\/cli(?!@)/`. This reliably distinguishes `@camunda8/cli` (stable) from `@camunda8/cli@alpha` regardless of surrounding characters.

File: tests/unit/update-check.test.ts [hunk 9]
```diff
@@ -0,0 +1,512 @@
...
+  test('alpha prerelease returns "alpha"', () => {
+    assert.strictEqual(detectChannel('1.2.3-alpha.5'), 'alpha');
+  });
+
+  test('other prerelease returns "latest"', () => {
+    assert.strictEqual(detectChannel('1.2.3-beta.1'), 'latest');
+  });
+
+  test('version with alpha in patch (not prerelease) returns "latest"', () => {
+    // Edge case: "alpha" in the version but not as a prerelease tag
```
[the remark this answers is not in the document]
Comment (jwulf) [reply tests/unit/update-check.test.ts]: Fixed in 30df361. Renamed the test to `'stable version returns "latest"'` since the input `'1.0.0'` doesn't contain "alpha" at all.

File: tests/unit/update-check.test.ts [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+    assert.ok(existsSync(cachePath), 'Cache file should be created');
+
+    const cache = JSON.parse(readFileSync(cachePath, 'utf-8'));
+    assert.strictEqual(cache.notifiedVersion, '2.0.0');
+  });
+
+  test('alpha channel user checks alpha dist-tag, not latest', async () => {
+    let requestedUrl = '';
+    globalThis.fetch = async (input: string | URL | Request) => {
+      requestedUrl = typeof input === 'string' ? input : input.toString();
```
[the remark this answers is not in the document]
Comment (jwulf) [reply tests/unit/update-check.test.ts]: Fixed in c85bcc3. Removed the unused `requestedUrl` variable and simplified the fetch mock signature.

File: src/update-check.ts [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+
+	if (!pendingNotification) return;
+
+	// Suppress in JSON mode — don't pollute structured output
+	if (c8ctl.outputMode === "json") return;
+
+	// Persist so we don't nag about this version again (deferred from background
+	// check so JSON-mode suppression doesn't write the cache prematurely)
+	writeCache({
+		...readCache(),
```
[the remark this answers is not in the document]
Comment (jwulf) [reply src/update-check.ts]: Fixed in c85bcc3. Added `?? {}` before spreading, consistent with the `persistPatientTimestamp()` pattern.

File: src/update-check.ts [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+			fetchController?.abort();
+			return;
+		}
+	}
+
+	if (!pendingNotification) return;
+
+	// Suppress in JSON mode — don't pollute structured output
+	if (c8ctl.outputMode === "json") return;
+
```
[the remark this answers is not in the document]
Comment (jwulf) [reply src/update-check.ts]: Fixed in c85bcc3. JSON mode check is now the first thing in `printUpdateNotification()` — it aborts any pending fetch and returns immediately, guaranteeing zero post-command delay for piped/automated usage.

File: tests/unit/update-check.test.ts [hunk 13]
```diff
@@ -0,0 +1,516 @@
...
+    assert.strictEqual(detectChannel('1.2.3-alpha.5'), 'alpha');
+  });
+
+  test('other prerelease returns "latest"', () => {
+    assert.strictEqual(detectChannel('1.2.3-beta.1'), 'latest');
+  });
+
+  test('stable version returns "latest"', () => {
+    assert.strictEqual(detectChannel('1.0.0'), 'latest');
+  });
```
[the remark this answers is not in the document]
Comment (jwulf) [reply tests/unit/update-check.test.ts:37]: Fixed in c85bcc3. Renamed to `'version containing alpha outside prerelease returns "latest"'` with input `'1.0.0-alphabeta.1'` — now tests a genuinely distinct edge case where "alpha" appears in the string but not as the `-alpha.N` prerelease pattern.
